datastore/redis.go

```diff
@@ -386,38 +393,119 @@ func (r *RedisCache) SaveLatestBuilderBid(slot uint64, builderPubkey, parentHash
+	bidValueIntMap := make(map[string]*big.Int)
 	if err != nil {
-		return err
+		return wasTopBidUpdated, topBidValue, topBidBuilderPubkey, err
 	}
 
-	// Find bid with highest value among all the latest bids
-	topBidValue := big.NewInt(0)
-	topBidBuilderPubkey := ""
+	// 2. Find bid with highest value among all the latest bids
 	for builderPubkey, bidValue := range bidValueMap {
 		val := new(big.Int)
 		val.SetString(bidValue, 10)
-		if val.Cmp(topBidValue) > 0 {
-			topBidValue = val
-			topBidBuilderPubkey = builderPubkey
+		bidValueIntMap[builderPubkey] = val
+		if val.Cmp(prevTopBidValue) > 0 {
+			prevTopBidValue = val
 		}
+
+		// Keep track of the top value of this builder, to discard any lower bids if not using cancellations
+		if builderPubkey == payload.BuilderPubkey().String() && val.Cmp(currentBuilderLastTopBidValue) > 0 {
+			currentBuilderLastTopBidValue = val
+		}
+	}
+
+	// 3. Check whether to continue at all
+	// - In cancellation mode: always save latest bid
+	// - In non-cancellation mode: abort if current bid is lower than this builders previous bid
+	//
+	// Example bid update orders (BA=non cancelling builder, BC=builder using cancellations):
+	//    BA1=5, BC1=10, BA2=6, BC2=5 -> winning bid should be BA2 (6)
+	//    BA1=5, BC1=10, BA2=4, BC2=5 -> winning bid should be BA1 (5)
+	if !isCancellationEnabled && payload.Value().Cmp(currentBuilderLastTopBidValue) < 1 {
+		return wasTopBidUpdated, prevTopBidValue, topBidBuilderPubkey, nil
 	}
 
-	if topBidBuilderPubkey == "" {
-		return ErrFailedUpdatingTopBidNoBids
+	// Time to save things in Redis
+	// 1. Save the execution payload
+	err = r.SaveExecutionPayload(payload.Slot(), payload.ProposerPubkey(), payload.BlockHash(), getPayloadResponse)
+	if err != nil {
+		return wasTopBidUpdated, prevTopBidValue, topBidBuilderPubkey, err
 	}
 
-	// Get the actual bid
-	keyBid := r.keyBlockBuilderLatestBids(slot, parentHash, proposerPubkey)
+	// 2. Save this bid
+	err = r.SaveBuilderBid(payload.Slot(), payload.BuilderPubkey().String(), payload.ParentHash(), payload.ProposerPubkey(), reqReceivedAt, getHeaderResponse)
+	if err != nil {
+		return wasTopBidUpdated, prevTopBidValue, topBidBuilderPubkey, err
+	}
+
+	//
+	// Decide if winning bid should be updated
+	//
+	if payload.Value().Cmp(prevTopBidValue) == 1 {
+		// update if this payload is new top bid
+		topBidValue = payload.Value()
+		topBidBuilderPubkey = payload.BuilderPubkey().String()
+	} else if isCancellationEnabled {
+		// In cancellation mode, builder can overwrite his own last bid with lower value.
+		// First, compute newest top bid with latest values
+		for builderPubkey, bidValue := range bidValueIntMap {
+			if builderPubkey == payload.BuilderPubkey().String() {
+				bidValue = payload.Value() // overwrite previous value with new value
+			}
+			if bidValue.Cmp(topBidValue) > 0 {
+				topBidValue = bidValue
+				topBidBuilderPubkey = builderPubkey
+			}
```

Comment:
There was [another suggestion](https://github.com/flashbots/mev-boost-relay/pull/350#discussion_r1167120054) to create a separate struct to encapsulate the return values

I'm wondering if it would help if we did something like:
```
type TopBidResponse struct { // note: for illustrative purposes only
	updated             bool
	prevTopBid          *big.Int
	currentTopBid       *big.Int
	TopBidBuilderPubKey string
}

func (tbr *TopBidResponse) tryUpdateTopBid(input *big.Int) (*big.Int, bool) {
        const IsGreater = 1

	switch input.Cmp(tbr.currentTopBid) {
	case IsGreater:
		tbr.prevTopBid, tbr.currentTopBid = tbr.currentTopBid, input
		tbr.updated = true
	default:
		break
	}
	
	return tbr.prevTopBid, tbr.updated
}

func (tbr *TopBidResponse) TryUpdateTopBid(payload *common.BuilderSubmitBlockRequest, cancel bool, bidMap map[string]*big.Int) {
	payloadVal := payload.Value()
	if _, updated := tbr.tryUpdateTopBid(payloadVal); updated {
		return
	}

	if !cancel {
		return
	}

	for builderPubKey, bidValue := range bidMap {
		if builderPubKey == payload.BuilderPubkey().String() {
			bidValue.Set(payloadVal)
		}

		tbr.tryUpdateTopBid(bidValue)
	}
}

func (tbr *TopBidResponse) WasUpdated() bool {
	return tbr.updated
}
```

then within `SaveBidAndUpdateTopBid`, something like:
```
topBidResponse := &TopBidResponse{// TODO: Complete Me}
topBidResponse.TryUpdateTopBid(payload, isCancellationEnabled, bidValueIntMap)
if !topBidResponse.WasUpdated() {
  return topBidResponse, nil
} 
```

Since we'd be instantiating `TopBidResponse` based off the scope of the current request, we shouldn't need to read from Redis again if i'm understanding this correctly

Reply:
yeah the struct method was more as an example - i think it might be worth creating something like a `BidUpdater` interface (the name doesn't have to be BidUpdater) - imo the redis cache methods should hold cache specific logic, but atm it holds "bid specific" logic. I think it's worth separating those concerns so that it's easier for us to reason about and maintain

Reply:
I hear you. But I am wary of making too big changes, I'm more more looking for an incremental refactoring here right now. 

Also, the top bid specific logic is tied deeply into redis, which is the whole source of truth for this.

Reply:
Basically, updating a bd and computing the new top bid is just a bunch of redis operations

Reply:
okay, did a big refactoring taking all this into account! take a look https://github.com/flashbots/mev-boost-relay/blob/submitblock-performance-1/datastore/redis.go#L409